src/app/modules/shared_modules/keycard_popup/module.nim

```diff
@@ -570,16 +557,32 @@ proc buildKeyPairItemBasedOnCardMetadata[T](self: Module[T], cardMetadata: CardM
     result.item.setKeyUid(currKp.getKeyUid())
     result.item.setPubKey(currKp.getPubKey())
   result.knownKeyPair = true
-  var unknonwAccountNumber = 0
-  for wa in cardMetadata.walletAccounts:
-    if self.updateKeyPairItemIfDataAreKnown(wa.address, result.item):
-      continue
-    let (balance, balanceFetched) = self.controller.getOrFetchBalanceForAddressInPreferredCurrency(wa.address)
-    result.knownKeyPair = false
-    unknonwAccountNumber.inc
-    let name = atc.KEYCARD_ACCOUNT_NAME_OF_UNKNOWN_WALLET_ACCOUNT & $unknonwAccountNumber
-    result.item.addAccount(newKeyPairAccountItem(name, wa.path, wa.address, pubKey = wa.publicKey, emoji = "",
-      colorId = "", icon = "undefined", balance, balanceFetched))
+  # resolve known accounts first
+  let accounts = self.controller.getWalletAccounts()
+  for acc in accounts:
+    for cardAcc in cardMetadata.walletAccounts:
+      if acc.isChat or acc.walletType == WalletTypeWatch or cmpIgnoreCase(acc.address, cardAcc.address) != 0:
+        continue
+      var icon = ""
+      if acc.emoji.len == 0:
+        icon = "wallet"
+      result.item.addAccount(newKeyPairAccountItem(acc.name, acc.path, acc.address, acc.publicKey, acc.emoji, acc.colorId, icon,
+        balance = 0.0, balanceFetched = true))
+  # handle unknown accounts
```

Comment:
how is that possible?
We specifically ask tokens data for the known account?

Reply:
Ahh that's in the context of keycard :)
Let's say your app and keycard has 2 accounts like acc1 and acc2, that means your accounts are added to the app and then added to a keycard as well, cause in order to add new account you have to authenticate that action and we silently sync the keycard state. Now you want to add acc3, that means you need to authenticate that action, once acc3 gets added to the app (because you've provided a pin) we sync the accounts on a keycard, means your keycard also knows about acc3.

But now let's say you want to delete acc2, for that action you won't be asked to authenticate it, means after account deletion your app knows for acc1 and acc3, but doesn't know about acc2. That's perfectly fine, but when you go to the keycard details view, you will see acc1 and acc3 with emoji, name and color you have set for them, but acc2 will be grey without emoji, notifying user that keycard state is not synced with the app. The first time you provide a pin, from any reason the state of that keycard will be synced with the app state. So that's about app <=> keycard syncing, not real device syncing.

Reply:
Should the fact your account is in keycard matters here, cause we have the list of account in memory and they come from keycard or not. We then ask for token for our list of account.

So it is necessarily known?